File: runtime/entity/tx/handlers/lending.ts
import type { AccountTx, EntityInput, EntityState, EntityTx } from '../../../types';
import { normalizeInterestBps, normalizeLendingTerm } from '../../../extensions/lending';
import { addMessage, cloneEntityState } from '../../../state-helpers';
import type { MempoolOp } from './account';

type EntityTxOf<T extends EntityTx['type']> = Extract<EntityTx, { type: T }>;

type LendingResult = {
  newState: EntityState;
  outputs: EntityInput[];
  mempoolOps?: MempoolOp[];
};

const INTENT_ID_RE = /^(?:lend|borrow|loan)-[0-9a-f]{16}$/;

const normalized = (value: unknown): string => String(value || '').trim().toLowerCase();

const processingTrigger = (state: EntityState): EntityInput[] => {
  const firstValidator = state.config.validators[0];
  return firstValidator
    ? [{ entityId: state.entityId, signerId: firstValidator, entityTxs: [] }]
    : [];
};

const requireIntentId = (value: string, prefix: 'lend' | 'borrow' | 'loan'): void => {
  const id = normalized(value);
  if (!INTENT_ID_RE.test(id) || !id.startsWith(`${prefix}-`)) {
    throw new Error(`LENDING_INTENT_ID_INVALID:${value}`);
  }
};

const requirePositiveAmount = (amount: bigint, context: string): void => {
  if (amount <= 0n) throw new Error(`${context}_AMOUNT_MUST_BE_POSITIVE`);
};

const requireHubAccount = (state: EntityState, hubEntityId: string): string => {
  const hub = normalized(hubEntityId);
  if (!hub || !state.accounts.has(hub)) {
    throw new Error(`LENDING_HUB_ACCOUNT_MISSING:${hub || 'missing'}`);
  }
  return hub;
};

const queueAccountTx = (
  state: EntityState,
  hubEntityId: string,
  tx: AccountTx,
  message: string,
): LendingResult => {
  const newState = cloneEntityState(state);
  addMessage(newState, message);
  return {
    newState,
    outputs: processingTrigger(state),
    mempoolOps: [{ accountId: hubEntityId, tx }],
  };
};

export const handleLendingOfferEntityTx = (
  entityState: EntityState,
  entityTx: EntityTxOf<'lendingOffer'>,
): LendingResult => {
  const hubEntityId = requireHubAccount(entityState, entityTx.data.hubEntityId);
  requireIntentId(entityTx.data.positionId, 'lend');
  requirePositiveAmount(entityTx.data.amount, 'LENDING_FUND');
  const termId = normalizeLendingTerm(entityTx.data.termId);
  const interestBps = normalizeInterestBps(entityTx.data.interestBps);
  const account = entityState.accounts.get(hubEntityId)!;
  if (!account.deltas.has(entityTx.data.tokenId)) {
    throw new Error(`LENDING_TOKEN_NOT_ENABLED:${entityTx.data.tokenId}`);
  }
  return queueAccountTx(entityState, hubEntityId, {
    type: 'lending_fund',
    data: {
      positionId: normalized(entityTx.data.positionId),
      hubEntityId,
      lenderEntityId: normalized(entityState.entityId),
      tokenId: entityTx.data.tokenId,
      amount: entityTx.data.amount,
      termId,
      interestBps,
    },
  }, `Lending pool funding requested: ${entityTx.data.amount} token=${entityTx.data.tokenId}`);
};

export const handleLendingBorrowEntityTx = (
  entityState: EntityState,
  entityTx: EntityTxOf<'lendingBorrow'>,
): LendingResult => {
  const hubEntityId = requireHubAccount(entityState, entityTx.data.hubEntityId);
  requireIntentId(entityTx.data.requestId, 'borrow');
  requirePositiveAmount(entityTx.data.amount, 'LENDING_BORROW');
  const termId = normalizeLendingTerm(entityTx.data.termId);
  const maxInterestBps = normalizeInterestBps(entityTx.data.maxInterestBps ?? 10_000);
  return queueAccountTx(entityState, hubEntityId, {
    type: 'lending_borrow_request',
    data: {
      requestId: normalized(entityTx.data.requestId),
      hubEntityId,
      borrowerEntityId: normalized(entityState.entityId),
      tokenId: entityTx.data.tokenId,
      amount: entityTx.data.amount,
      termId,
      maxInterestBps,
    },
  }, `Loan requested: ${entityTx.data.amount} token=${entityTx.data.tokenId}`);
};

export const handleLendingRepayEntityTx = (
  entityState: EntityState,
  entityTx: EntityTxOf<'lendingRepay'>,
): LendingResult => {
  const hubEntityId = requireHubAccount(entityState, entityTx.data.hubEntityId);
  requireIntentId(entityTx.data.loanId, 'loan');
  requirePositiveAmount(entityTx.data.amount, 'LENDING_REPAY');
  return queueAccountTx(entityState, hubEntityId, {
    type: 'lending_repay',
    data: {
      loanId: normalized(entityTx.data.loanId),
      hubEntityId,
      borrowerEntityId: normalized(entityState.entityId),
      tokenId: entityTx.data.tokenId,
      amount: entityTx.data.amount,
    },
  }, `Loan repayment requested: ${entityTx.data.loanId}`);
};

export const handleLendingClosePositionEntityTx = (
  entityState: EntityState,
  entityTx: EntityTxOf<'lendingClosePosition'>,
): LendingResult => {
  const hubEntityId = requireHubAccount(entityState, entityTx.data.hubEntityId);
  requireIntentId(entityTx.data.positionId, 'lend');
  return queueAccountTx(entityState, hubEntityId, {
    type: 'lending_close_request',
    data: {
      positionId: normalized(entityTx.data.positionId),
      hubEntityId,
      lenderEntityId: normalized(entityState.entityId),
    },
  }, `Lending position close requested: ${entityTx.data.positionId}`);
};

File: rscore/crates/entity-kernel/src/local_financial/lending.rs
use num_bigint::BigInt;
use xln_rscore_engine::AccountTx;

use crate::{EntityFrameEvent, EntityKernelError, EntityStateSlice};

use super::types::{
    LendingBorrowEntityTx, LendingClosePositionEntityTx, LendingOfferEntityTx,
    LendingRepayEntityTx, LocalAccountFinancialView,
};

fn valid_intent(value: &str, prefix: &str) -> bool {
    value
        .strip_prefix(prefix)
        .and_then(|value| value.strip_prefix('-'))
        .is_some_and(|suffix| {
            suffix.len() == 16 && suffix.bytes().all(|byte| byte.is_ascii_hexdigit())
        })
}

fn require_account(state: &EntityStateSlice, hub: &str) -> Result<(), EntityKernelError> {
    if !state.known_accounts.contains(hub) {
        return Err(EntityKernelError::lending("HUB_ACCOUNT_MISSING"));
    }
    Ok(())
}

fn queue(
    state: &EntityStateSlice,
    hub: String,
    tx: AccountTx,
    message: String,
    account_txs: &mut Vec<(String, AccountTx)>,
    events: &mut Vec<EntityFrameEvent>,
    wakes: &mut Vec<String>,
) {
    account_txs.push((hub, tx));
    events.push(EntityFrameEvent::Status { message });
    wakes.push(state.entity_id.clone());
}

pub(super) fn apply_offer(
    state: &EntityStateSlice,
    tx: LendingOfferEntityTx,
    views: &std::collections::BTreeMap<String, LocalAccountFinancialView>,
    account_txs: &mut Vec<(String, AccountTx)>,
    events: &mut Vec<EntityFrameEvent>,
    wakes: &mut Vec<String>,
) -> Result<(), EntityKernelError> {
    require_account(state, &tx.hub_entity_id)?;
    let view = views
        .get(&tx.hub_entity_id)
        .ok_or_else(|| EntityKernelError::lending("ACCOUNT_VIEW_MISSING"))?;
    if !valid_intent(&tx.position_id, "lend") {
        return Err(EntityKernelError::lending("INTENT_ID_INVALID"));
    }
    if tx.amount <= BigInt::from(0) {
        return Err(EntityKernelError::lending("FUND_AMOUNT_MUST_BE_POSITIVE"));
    }
    if !view.owner_out_capacity.contains_key(&tx.token_id) {
        return Err(EntityKernelError::lending("TOKEN_NOT_ENABLED"));
    }
    let hub = tx.hub_entity_id.clone();
    queue(
        state,
        hub.clone(),
        AccountTx::LendingFund {
            position_id: tx.position_id,
            hub_entity_id: hub,
            lender_entity_id: state.entity_id.clone(),
            token_id: tx.token_id,
            amount: tx.amount.clone(),
            term_id: tx.term_id,
            interest_bps: i64::from(tx.interest_bps),
        },
        format!(
            "Lending pool funding requested: {} token={}",
            tx.amount,
            tx.token_id.get()
        ),
        account_txs,
        events,
        wakes,
    );
    Ok(())
}

pub(super) fn apply_borrow(
    state: &EntityStateSlice,
    tx: LendingBorrowEntityTx,
    account_txs: &mut Vec<(String, AccountTx)>,
    events: &mut Vec<EntityFrameEvent>,
    wakes: &mut Vec<String>,
) -> Result<(), EntityKernelError> {
    require_account(state, &tx.hub_entity_id)?;
    if !valid_intent(&tx.request_id, "borrow") {
        return Err(EntityKernelError::lending("INTENT_ID_INVALID"));
    }
    if tx.amount <= BigInt::from(0) {
        return Err(EntityKernelError::lending("BORROW_AMOUNT_MUST_BE_POSITIVE"));
    }
    let hub = tx.hub_entity_id.clone();
    queue(
        state,
        hub.clone(),
        AccountTx::LendingBorrowRequest {
            request_id: tx.request_id,
            hub_entity_id: hub,
            borrower_entity_id: state.entity_id.clone(),
            token_id: tx.token_id,
            amount: tx.amount.clone(),
            term_id: tx.term_id,
            max_interest_bps: i64::from(tx.max_interest_bps),
        },
        format!("Loan requested: {} token={}", tx.amount, tx.token_id),
        account_txs,
        events,
        wakes,
    );
    Ok(())
}

pub(super) fn apply_repay(
    state: &EntityStateSlice,
    tx: LendingRepayEntityTx,
    account_txs: &mut Vec<(String, AccountTx)>,
    events: &mut Vec<EntityFrameEvent>,
    wakes: &mut Vec<String>,
) -> Result<(), EntityKernelError> {
    require_account(state, &tx.hub_entity_id)?;
    if !valid_intent(&tx.loan_id, "loan") {
        return Err(EntityKernelError::lending("INTENT_ID_INVALID"));
    }
    if tx.amount <= BigInt::from(0) {
        return Err(EntityKernelError::lending("REPAY_AMOUNT_MUST_BE_POSITIVE"));
    }
    let hub = tx.hub_entity_id.clone();
    let loan_id = tx.loan_id.clone();
    queue(
        state,
        hub.clone(),
        AccountTx::LendingRepay {
            loan_id: tx.loan_id,
            hub_entity_id: hub,
            borrower_entity_id: state.entity_id.clone(),
            token_id: tx.token_id,
            amount: tx.amount,
        },
        format!("Loan repayment requested: {loan_id}"),
        account_txs,
        events,
        wakes,
    );
    Ok(())
}

pub(super) fn apply_close(
    state: &EntityStateSlice,
    tx: LendingClosePositionEntityTx,
    account_txs: &mut Vec<(String, AccountTx)>,
    events: &mut Vec<EntityFrameEvent>,
    wakes: &mut Vec<String>,
) -> Result<(), EntityKernelError> {
    require_account(state, &tx.hub_entity_id)?;
    if !valid_intent(&tx.position_id, "lend") {
        return Err(EntityKernelError::lending("INTENT_ID_INVALID"));
    }
    let hub = tx.hub_entity_id.clone();
    let position_id = tx.position_id.clone();
    queue(
        state,
        hub.clone(),
        AccountTx::LendingCloseRequest {
            position_id: tx.position_id,
            hub_entity_id: hub,
            lender_entity_id: state.entity_id.clone(),
        },
        format!("Lending position close requested: {position_id}"),
        account_txs,
        events,
        wakes,
    );
    Ok(())
}

#[cfg(test)]
mod tests {
    use std::collections::BTreeMap;

    use xln_rscore_engine::{LendingTermId, TokenId};

    use super::*;

    fn entity(byte: &str) -> String {
        format!("0x{}", byte.repeat(32))
    }

    #[test]
    fn offer_projects_the_exact_typescript_account_transaction_and_wake() {
        let owner = entity("20");
        let hub = entity("10");
        let token = TokenId::new(1).expect("token");
        let mut state = EntityStateSlice::empty(owner.clone(), 1_000);
        state.known_accounts.insert(hub.clone());
        let views = BTreeMap::from([(
            hub.clone(),
            LocalAccountFinancialView {
                active: true,
                owner_side: xln_rscore_engine::Side::Left,
                owner_out_capacity: BTreeMap::from([(token, BigInt::from(20_000))]),
                owner_peer_credit_limit: BTreeMap::new(),
                settlement_workspace: None,
                settlement_transition_pending: false,
                settlement_execution: Err("SETTLEMENT_WORKSPACE_MISSING".into()),
                rebalance_active_quote: None,
                htlc_locks: BTreeMap::new(),
                pulls: BTreeMap::new(),
                swap_offers: BTreeMap::new(),
                pending_cross_pull_close_ids: Default::default(),
                pending_cross_swap_ack_ids: Default::default(),
                dispute: None,
            },
        )]);
        let (mut account_txs, mut events, mut wakes) = (Vec::new(), Vec::new(), Vec::new());
        apply_offer(
            &state,
            LendingOfferEntityTx {
                position_id: "lend-1111111111111111".into(),
                hub_entity_id: hub.clone(),
                token_id: token,
                amount: BigInt::from(1_000),
                term_id: LendingTermId::OneDay,
                interest_bps: 100,
            },
            &views,
            &mut account_txs,
            &mut events,
            &mut wakes,
        )
        .expect("offer");
        assert!(matches!(
            account_txs.as_slice(),
            [(target, AccountTx::LendingFund {
                position_id,
                lender_entity_id,
                amount,
                ..
            })] if target == &hub
                && position_id == "lend-1111111111111111"
                && lender_entity_id == &owner
                && amount == &BigInt::from(1_000)
        ));
        assert_eq!(wakes, vec![owner]);
        assert!(matches!(
            events.as_slice(),
            [EntityFrameEvent::Status { message }]
                if message == "Lending pool funding requested: 1000 token=1"
        ));
    }
}
